genrl/classical/bandit/contextual_policies.py

```diff
@@ -0,0 +1,381 @@
+
+    :param bandit: The Bandit to solve
+    :param c: Confidence level which controls degree of exploration 
+    :type bandit: ContextualBandit type object 
+    :type c: float 
+    """
+
+    def __init__(self, bandit: ContextualBandit, c: float = 1.0):
+        super(UCBCBPolicy, self).__init__(bandit, requires_init_run=True)
+        self._c = c
+        self._Q = np.zeros(shape=(bandit.bandits, bandit.arms))
+        self._counts = np.zeros(shape=(bandit.bandits, bandit.arms))
+
+    @property
+    def c(self) -> float:
+        """
+        Get the confidence level which weights the exploration term 
+        
+        :returns: Confidence level which controls degree of exploration 
+        :rtype: float
+        """
+        return self._c
+
+    @property
+    def Q(self) -> np.ndarray:
+        """
+        Get the q values assigned by the policy to all actions
+
+        :returns: Numpy array of q values for all actions
+        :rtype: numpy.ndarray
+        """
+        return self._Q
+
+    @property
+    def counts(self) -> np.ndarray:
+        """
+        Get the number of times each action has been taken
+
+        :returns: Numpy array with count for each action
+        :rtype: numpy.ndarray
+        """
+        return self._counts
+
+    def select_action(self, context: int, t: int) -> int:
+        """
+        Select an action according to upper confidence bound action selction
+
+        Take action that maximises a weighted sum of the Q values for the action
+        and an exploration encouragement term controlled by c.
+
+        :param context: the context to select action for
+        :param t: timestep to choose action for
+        :type context: int
+        :type t: int
+        :returns: Selected action
+        :rtype: int
+        """
+        action = np.argmax(
+            self.Q[context]
+            + self.c * np.sqrt(2 * np.log(t + 1) / (self.counts[context] + 1))
+        )
+        self.action_hist.append(action)
+        return action
+
+    def update_params(self, context: int, action: int, reward: float) -> None:
+        """
+        Update parmeters for the policy
+
+        Updates the regret as the difference between max Q value and 
+        that of the action. Updates the Q values according to the
+        reward recieved in this step.
+
+        :param context: context for which action is taken
+        :param action: action taken for the step
+        :param reward: reward obtained for the step
+        :type context: int
+        :type action: int
+        :type reward: float
+        """
+        self.reward_hist.append(reward)
+        self._regret += max(self.Q[context]) - self.Q[context, action]
+        self.regret_hist.append(self.regret)
+        self.Q[context, action] += (reward - self.Q[context, action]) / (
+            self.counts[context, action] + 1
+        )
+        self.counts[context, action] += 1
+
+
+if __name__ == "__main__":
```

Comment:
Everything after this can be shifted to a separate file in the directory `examples`.